### PR TITLE
Add support for checking of email existence through the `signin` function

### PR DESCRIPTION
**Pull request by SudharakaP**
Related to https://github.com/opencollective/opencollective/issues/5572

Related Front-end PR: https://github.com/opencollective/opencollective-frontend/pull/7832

### CODE REVIEW

**Comment by znarf** [comment]
Last time we discussed this, we said this was not needed in this version, what changed?

**Comment by SudharakaP** [comment]
> Last time we discussed this, we said this was not needed in this version, what changed?

Sorry I was a bit confused on this. I've changed this PR so that we are now using the same `signin` function for the email existence check.

File: server/controllers/users.js
Comment (znarf) [inline server/controllers/users.js:56]: In which case do we have createProfile set to TRUE, set to FALSE?

File: server/controllers/users.js
Comment (znarf) [inline server/controllers/users.js:66]: Was that tested ?

File: server/controllers/users.js
Comment (znarf) [reply server/controllers/users.js:56]: We should add a code comment and delete this behavior in the future as it's not necessary.

Comment (SudharakaP) [reply server/controllers/users.js:60]: I've added a comment. This function needs to be refactored after we fully remove the `signin.js` on front-end (remove the old signin flow).

File: server/controllers/users.js
Comment (SudharakaP) [reply server/controllers/users.js]: This `else if` never triggered currently from the front-end. The reason is whenever we use `signin` function, we first call `exists` to see the email exists. Only if it exists we call the `signin` function. However I checked if this works correctly with a mock (calling the `signin` function regardless of whether email exists or not). We could leave this for now and once the old signin flow is removed from front-end could refactor this function to only sign-in and remove this collective creation logic. 🤔

Comment (SudharakaP) [reply server/controllers/users.js]: Yes you are correct, I initially thought the accounts were created using this in the sign-in flow. But then realized that we call a different mutation to create the profiles and this branch wasn't hit at all which is why I initially missed testing this. Sorry about that. 😢

File: server/controllers/users.js
Comment (SudharakaP) [reply server/controllers/users.js:78]: Ops, you are right. I've added the catch here. 👍🏽

File: server/controllers/users.js
Comment (SudharakaP) [reply server/controllers/users.js]: Here, I am not returning an HTTP error code. Just the response with an error object. And on the front-end we check for the `EMAIL_DOES_NOT_EXIST` and if found show the appropriate page. Since this error is specific to open collective internal logic, I didn't see the need to use an HTTP error code; in essence the request was successfully processed; and we return the fact that the email is not found. Do you think we should use an HTTP error code? 🤔

File: server/controllers/users.js
Comment (SudharakaP) [reply server/controllers/users.js]: I've renamed this one to `user` and refactored.

File: server/controllers/users.js
Comment (Betree) [reply server/controllers/users.js]: GraphQL gives us the habit of not using HTTP error codes, but in the old days of the Internet that would probably have been a 404 - resource not found :-) @znarf wdyt?

Comment (znarf) [reply server/controllers/users.js]: 400 should do the job, I would not do 404 unless the URL is non existing

Comment (SudharakaP) [reply server/controllers/users.js]: @Betree : So you think we should return an HTTP error code ? My understanding is that Apollo docs discourage this and rather expects to send a custom error code which I have done; Reference: https://www.apollographql.com/docs/apollo-server/data/errors/#returning-http-status-codes

> We recommend using the included [Error Codes](https://www.apollographql.com/docs/apollo-server/data/errors/#error-codes) or [Custom Errors](https://www.apollographql.com/docs/apollo-server/data/errors/#custom-errors) for error consistency rather than directly modifying the HTTP response.

Comment (Betree) [reply server/controllers/users.js]: @SudharakaP We're not doing Apollo here, it's a REST endpoint - not GraphQL

Comment (SudharakaP) [reply server/controllers/users.js]: Ops sorry, I had some confusion here; I've added the HTTP 400 code and corrected the front-end as well;

https://github.com/opencollective/opencollective-frontend/pull/7832/commits/f3f0a28ef1ec42dfdcb93f1f10e2255d3d1dd272

**Review by znarf — CHANGES_REQUESTED** [review]
E2E tests are broken

**Comment by SudharakaP** [comment]
> E2E tests are broken

Sorry, made a mistake in a recent commit where I was using the wrong `user` object.

https://github.com/opencollective/opencollective-api/blob/48a3aac9b8f6a6392f87f3ff20700b0a1e42ac44/server/controllers/users.js#L69

I've corrected this, the tests should now be passing. 

There's a codeql failure, but that doesn't seem to be related to this PR; https://github.com/opencollective/opencollective-api/security/code-scanning/51 and is about rate-limiting the `signin` endpoint. Do you think we have to implement this, or my thinking is that we could use the [Cloudflare advance rate limiting](https://www.cloudflare.com/rate-limiting/) to tackle this? 🤔

**Comment by znarf** [comment]
It's a good idea to have rate limiting, we have many examples in our source code and there is no need to rely on Cloudflare for it.

**Comment by znarf** [comment]
Check the previous `exists` behavior for rate limiting.

**Comment by SudharakaP** [comment]
> Check the previous `exists` behavior for rate limiting.

Thanks. If it's okay, we'll extract this to a separate issue; https://github.com/opencollective/opencollective/issues/5630 so that we are not blocking this PR. 🤔

**Comment by znarf** [comment]
@SudharakaP we can't ship this in production without rate limiting

**Comment by SudharakaP** [comment]
> @SudharakaP we can't ship this in production without rate limiting

Sure, I'll address this today, I thought this codeql issue wasn't related initially as we were using the `/signin` endpoint and there was no rate limiting anyways. 🤔 

Anyways, I've pushed a commit to this PR with the rate limiting part. Let me know how it looks. 

image

File: config/default.json
Comment (SudharakaP) [inline config/default.json:71]: @znarf : I wasn't exactly sure on this limit here; do you think 50 is reasonable here? or should we reduce it to less. My gut feeling initially was that something like 20 would be better but wasn't sure if it's too aggressive. 🤔

**Review by Betree — APPROVED** [review]
The code looks good to me. Let's wait for a final from @znarf too before merging

**Review by znarf — APPROVED** [review]

**Comment by SudharakaP** [comment]
@znarf @Betree : I am a bit perplexed looking at the CodeQL results and it seems that it still fails with https://github.com/opencollective/opencollective-api/security/code-scanning/51 although we have now implemented the rate limiting within `signin`. My thinking is that GitHub isn't picking up the fact we have implemented it within the function but rather wants a solution that applies to all requests. Shall we dismiss this alert? 🤔

**Comment by znarf** [comment]
To be clear, we didn't do the rate limiting to make CodeQL "happy", we did it because it was important and a regression compared to previous behavior.

**Pull request merged**
